// futu-api@1.2.9/dist/src/futu.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const subscribe_1 = __importDefault(require("./decorators/subscribe"));
exports.Subscribe = subscribe_1.default;
const Proto = __importStar(require("./proto/proto"));
exports.Proto = Proto;
const ws_1 = __importDefault(require("./ws"));
const DefaultConfig = {
    ip: 'localhost',
    port: 33333,
    recvNotify: true,
    accMarket: Proto.Trd_Common.TrdMarket.TrdMarket_HK,
    accEnv: Proto.Trd_Common.TrdEnv.TrdEnv_Real,
    accType: Proto.Trd_Common.TrdAccType.TrdAccType_Cash,
    isSSL: false,
    wsKey: null,
    reqTimeout: 10000
};
class Futu extends ws_1.default {
    constructor(config, onOpenListener) {
        super(Object.assign({}, DefaultConfig, config), () => onOpenListener === null || onOpenListener === void 0 ? void 0 : onOpenListener(this));
    }
    get ready() {
        return super.ready;
    }
    close() {
        super.close();
    }
    async getGlobalState(req) {
        return super.request('GetGlobalState', req);
    }
    async notify() {
        return super.request('Notify');
    }
    async keepAlive(req) {
        return super.request('KeepAlive', req);
    }
    async getUserInfo(req) {
        return super.request('GetUserInfo', req);
    }
    async getDelayStatistics(req) {
        return super.request('GetDelayStatistics', req);
    }
    async qotSub(req) {
        return super.request('Qot_Sub', req);
    }
    async qotRegQotPush(req) {
        return super.request('Qot_RegQotPush', req);
    }
    async qotGetSubInfo(req) {
        return super.request('Qot_GetSubInfo', req);
    }
    async qotGetBasicQot(req) {
        return super.request('Qot_GetBasicQot', req);
    }
    async qotUpdateBasicQot() {
        return super.request('Qot_UpdateBasicQot');
    }
    async qotGetKL(req) {
        return super.request('Qot_GetKL', req);
    }
    async qotUpdateKL() {
        return super.request('Qot_UpdateKL');
    }
    async qotGetRT(req) {
        return super.request('Qot_GetRT', req);
    }
    async qotUpdateRT() {
        return super.request('Qot_UpdateRT');
    }
    async qotGetTicker(req) {
        return super.request('Qot_GetTicker', req);
    }
    async qotUpdateTicker() {
        return super.request('Qot_UpdateTicker');
    }
    async qotGetOrderBook(req) {
        return super.request('Qot_GetOrderBook', req);
    }
    async qotUpdateOrderBook() {
        return super.request('Qot_UpdateOrderBook');
    }
    async qotGetBroker(req) {
        return super.request('Qot_GetBroker', req);
    }
    async qotUpdateBroker() {
        return super.request('Qot_UpdateBroker');
    }
    async qotGetOrderDetail(req) {
        return super.request('Qot_GetOrderDetail', req);
    }
    async qotUpdateOrderDetail() {
        return super.request('Qot_UpdateOrderDetail');
    }
    async qotUpdatePriceReminder() {
        return super.request('Qot_UpdatePriceReminder');
    }
    async qotGetHistoryKL(req) {
        return super.request('Qot_GetHistoryKL', req);
    }
    async qotGetHistoryKLPoints(req) {
        return super.request('Qot_GetHistoryKLPoints', req);
    }
    async qotGetRehab(req) {
        return super.request('Qot_GetRehab', req);
    }
    async qotRequestHistoryKL(req) {
        return super.request('Qot_RequestHistoryKL', req);
    }
    async qotRequestHistoryKLQuota(req) {
        return super.request('Qot_RequestHistoryKLQuota', req);
    }
    async qotRequestRehab(req) {
        return super.request('Qot_RequestRehab', req);
    }
    async qotGetTradeDate(req) {
        return super.request('Qot_GetTradeDate', req);
    }
    async qotGetSuspend(req) {
        return super.request('Qot_GetSuspend', req);
    }
    async qotGetStaticInfo(req) {
        return super.request('Qot_GetStaticInfo', req);
    }
    async qotGetSecuritySnapshot(req) {
        return super.request('Qot_GetSecuritySnapshot', req);
    }
    async qotGetPlateSet(req) {
        return super.request('Qot_GetPlateSet', req);
    }
    async qotGetPlateSecurity(req) {
        return super.request('Qot_GetPlateSecurity', req);
    }
    async qotGetReference(req) {
        return super.request('Qot_GetReference', req);
    }
    async qotGetOwnerPlate(req) {
        return super.request('Qot_GetOwnerPlate', req);
    }
    async qotGetHoldingChangeList(req) {
        return super.request('Qot_GetHoldingChangeList', req);
    }
    async qotGetOptionChain(req) {
        return super.request('Qot_GetOptionChain', req);
    }
    async qotGetWarrant(req) {
        return super.request('Qot_GetWarrant', req);
    }
    async qotGetCapitalFlow(req) {
        return super.request('Qot_GetCapitalFlow', req);
    }
    async qotGetCapitalDistribution(req) {
        return super.request('Qot_GetCapitalDistribution', req);
    }
    async qotGetUserSecurity(req) {
        return super.request('Qot_GetUserSecurity', req);
    }
    async qotModifyUserSecurity(req) {
        return super.request('Qot_ModifyUserSecurity', req);
    }
    async qotStockFilter(req) {
        return super.request('Qot_StockFilter', req);
    }
    async qotGetCodeChange(req) {
        return super.request('Qot_GetCodeChange', req);
    }
    async qotGetIpoList(req) {
        return super.request('Qot_GetIpoList', req);
    }
    async qotGetFutureInfo(req) {
        return super.request('Qot_GetFutureInfo', req);
    }
    async qotRequestTradeDate(req) {
        return super.request('Qot_RequestTradeDate', req);
    }
    async qotSetPriceReminder(req) {
        return super.request('Qot_SetPriceReminder', req);
    }
    async qotGetPriceReminder(req) {
        return super.request('Qot_GetPriceReminder', req);
    }
    async qotGetUserSecurityGroup(req) {
        return super.request('Qot_GetUserSecurityGroup', req);
    }
    async qotGetMarketState(req) {
        return super.request('Qot_GetMarketState', req);
    }
    async trdGetAccList(req) {
        return super.request('Trd_GetAccList', req);
    }
    async trdUnlockTrade(req) {
        return super.request('Trd_UnlockTrade', req);
    }
    async trdSubAccPush(req) {
        return super.request('Trd_SubAccPush', req);
    }
    async trdGetFunds(req) {
        return super.request('Trd_GetFunds', req);
    }
    async trdGetPositionList(req) {
        return super.request('Trd_GetPositionList', req);
    }
    async trdGetMaxTrdQtys(req) {
        return super.request('Trd_GetMaxTrdQtys', req);
    }
    async trdGetOrderList(req) {
        return super.request('Trd_GetOrderList', req);
    }
    async trdPlaceOrder(req) {
        return super.request('Trd_PlaceOrder', req);
    }
    async trdModifyOrder(req) {
        return super.request('Trd_ModifyOrder', req);
    }
    async trdUpdateOrder() {
        return super.request('Trd_UpdateOrder');
    }
    async trdGetOrderFillList(req) {
        return super.request('Trd_GetOrderFillList', req);
    }
    async trdUpdateOrderFill() {
        return super.request('Trd_UpdateOrderFill');
    }
    async trdGetHistoryOrderList(req) {
        return super.request('Trd_GetHistoryOrderList', req);
    }
    async trdGetHistoryOrderFillList(req) {
        return super.request('Trd_GetHistoryOrderFillList', req);
    }
}
exports.default = Futu;
