// futu-api@1.2.9/dist/src/utils/init_promise.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class InitPromise {
    constructor() {
        this.isInit = false;
        this.resolveFunc = [];
        this.rejectFunc = [];
    }
    reset() {
        this.isInit = false;
        try {
            this.rejectFunc.forEach(reject => reject('reset'));
        }
        catch (_a) { }
        this.resolveFunc = [];
        this.rejectFunc = [];
    }
    get isReady() {
        if (this.isInit)
            return Promise.resolve();
        else
            return new Promise((resolve, reject) => {
                let timeoutTimer = setTimeout(() => reject('timeout'), 5000);
                this.resolveFunc.push(() => {
                    clearTimeout(timeoutTimer);
                    resolve();
                });
                this.rejectFunc.push((any) => reject(any));
            });
    }
    async resolve() {
        this.isInit = true;
        try {
            this.resolveFunc.forEach(resolve => resolve());
        }
        catch (_a) { }
        this.resolveFunc = [];
        this.rejectFunc = [];
    }
    async reject(_any) {
        this.isInit = false;
        try {
            this.rejectFunc.forEach(reject => reject(_any));
        }
        catch (_a) { }
        this.resolveFunc = [];
        this.rejectFunc = [];
    }
}
exports.default = InitPromise;
